### PR TITLE
Userland: Add an implementation of 'entr'

### PR DESCRIPTION
**Pull request by alimpfard**
This is like inotifywait, but it runs commands instead of printing
events out.

cc @nico - you suggested this; how would you suggest I word the commit?

### CODE REVIEW

**Comment by nico** [comment]
Hm, I'd probably put in an example,

```
To run a shell script every time it's saved:
  echo foo.sh | entr 'Shell foo.sh'

This is similar to
  while inotifywait -e modify -r foo.sh; do Shell foo.sh; done

but that misses events that happen during the body of the while loop. That can be fixed with

  inotifywait -m -e modify | while read dir action file; do Shell foo.sh; done

but that's harder to remember than the `entr` version.
```

Actually, maybe throw in a man page with some examples? (Or, if you don't like writing man pages, I can write on once this is in.)

File: Userland/entr.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+                mutable_old_pid.clear();
+            }
+
+            Vector<const char*> argv;
+            argv.append(executable_name);
+            for (auto arg : str_arguments)
+                argv.append(arg.characters());
+            argv.append(nullptr);
+
+            posix_spawn_file_actions_t action;
```
Comment (nico) [inline Userland/entr.cpp:152]: Doesn't look like you do anything with this other than create and destroy it -- can you just pass nullptr for the action instead?

**Comment by alimpfard** [comment]
Since you're offering, go ahead and write the manpage :P 
good idea with the example in the commit, I shall do that

File: Userland/entr.cpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+            started_job_ids.append(pid);
+            mutable_old_pid = pid;
+            if (verbose)
+                warnln("[exec] created instance for {} with pid={}", path, pid);
+        };
+
+        notifiers.empend(move(notifier));
+    }
+
+    Core::EventLoop::register_signal(SIGINT, [&](auto) {
```
Comment (nico) [inline Userland/entr.cpp:169]: Out of interest, why is this needed? Shouldn't the default SIGINT handler do something similar?

File: Userland/entr.cpp [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            notifier.set_enabled(false);
+            close(notifier.fd());
+            notifier.close();
+        }
+        _exit(0);
+    });
+
+    Core::EventLoop::register_signal(SIGCHLD, [&](auto) {
+        Vector<pid_t> new_pids;
+        for (auto& pid : started_job_ids) {
```
Comment (nico) [inline Userland/entr.cpp:180]: I think this handles coalesced SIGCHLDs correctly, but it feels "subtly correct" to me :) (No concrete suggestion here.)

File: Userland/entr.cpp (hunk 1 above)
Comment (alimpfard) [reply Userland/entr.cpp:152]: can I? I'm not too familiar with `posix_spawn()`, its manpage always talks about "the object pointed to by `attrp`" so I don't think it can be null?

File: Userland/entr.cpp (hunk 2 above)
Comment (alimpfard) [reply Userland/entr.cpp:169]: I guess, does it hurt to be sure?

File: Userland/entr.cpp (hunk 1 above)
Comment (nico) [reply Userland/entr.cpp:152]: Both Serenity's man page and https://pubs.opengroup.org/onlinepubs/009695399/functions/posix_spawn.html are fairly explicit that it can be nullptr.

File: Userland/entr.cpp (hunk 2 above)
Comment (nico) [reply Userland/entr.cpp:169]: If process teardown does this for you (and if it didn't, wouldn't it be a bug), why do it manually?

File: Userland/entr.cpp (hunk 1 above)
Comment (alimpfard) [reply Userland/entr.cpp]: I is bad at reading manpages, apparently.

**Comment by awesomekling** [comment]
I can't understand this commit message, as it assumes I know exactly how inotifywait works. Which I don't :)

**Comment by alimpfard** [comment]
oh this PR is still open? ^^'
let's close it for now, I remember finding some big holes in this idea and I don't feel like messing with it right now.

File: Userland/entr.cpp (hunk 3 above)
Comment (tomuta) [reply Userland/entr.cpp:168]: I know it's closed now, but for future reference: `WNOWAIT` is essentially just polling the information without un-waiting the child. So I'd expect zombies to hang around.

Comment (alimpfard) [reply Userland/entr.cpp:168]: L142 should take care of that though?
IIRC the main problem I saw with this was the fact that it wouldn't notice a change while it's running something

Comment (tomuta) [reply Userland/entr.cpp:168]: Possible, too complex logic for me to to comprehend ;) Also, I believe `WNOWAIT` will block if the process doesn't have waitable information (e.g. if it's still running) unless `WNOHANG` is specified. Not sure if that's intended since it's called in a loop.

Comment (alimpfard) [reply Userland/entr.cpp:168]: Ah whatever, the future me that decides to reopen this PR will think about it :shrug:
